File: examples/rust/simple_stat.rs
extern crate pmu;

use pmu::{Builder, CounterKind};

fn fib(n: usize) -> usize {
    let mut a = 1;
    let mut b = 1;

    for _ in 1..n {
        let old = a;
        a = b;
        b += old;
    }
    b
}

fn main() {
    let args: Vec<String> = std::env::args().collect();

    if args.len() != 2 {
        println!("Usage: simple_stat <n>");
        return;
    }

    let n = args[1].parse::<usize>().unwrap();

    let mut builder = Builder::new();
    builder.add_counter(CounterKind::Cycles);
    builder.add_counter(CounterKind::Instructions);
    builder.add_counter(CounterKind::Branches);
    builder.add_counter(CounterKind::BranchMisses);
    builder.add_counter(CounterKind::CacheHits);
    builder.add_counter(CounterKind::CacheMisses);

    let events = pmu::list_events();

    events
        .first()
        .and_then(|e| Some(builder.add_counter(CounterKind::System(e.clone()))));

    let uops = events
        .iter()
        .find(|e| e.to_string() == "HW:RETIRED_UOPS" || e.to_string() == "HW:UOPS_RETIRED.SLOTS");
    uops.and_then(|e| Some(builder.add_counter(CounterKind::System(e.clone()))));

    let mut counters = builder.build().unwrap();

    counters.start();
    let f = fib(n);
    counters.stop();

    println!("Fibonacci for {} is {}", n, f);

    for c in counters.iter() {
        println!("{} is {}", c.kind.to_string(), c.value);
    }
}

File: examples/cpp/simple_stat.cpp
#include <pmu/pmu.hpp>

#include <cstddef>
#include <cstdlib>
#include <iostream>

size_t fib(size_t n) {
  size_t a = 1;
  size_t b = 1;

  for (size_t i = 1; i < n; i++) {
    size_t old = a;
    a = b;
    b += old;
  }

  return b;
}

int main(int argc, char *argv[]) {
  int n = std::atoi(argv[1]);

  pmu::Builder builder;

  builder.add_counter(pmu::CounterKind::Cycles)
      .add_counter(pmu::CounterKind::Instructions)
      .add_counter(pmu::CounterKind::Branches)
      .add_counter(pmu::CacheCounter{pmu::CacheLevelKind::L1D,
                                     pmu::CacheCounterKind::Miss,
                                     pmu::CacheOpKind::Read})
      .add_counter(pmu::CounterKind::BranchMisses);

  auto counters = builder.build();

  counters->start();
  size_t f = fib(n);
  counters->stop();

  std::cout << "Fibonacci for " << n << " is " << f << "\n";

  for (auto value : *counters) {
    std::cout << value.name << ": " << value.value << "\n";
  }

  return 0;
}
